pkg/agent/qrm-plugins/memory/dynamicpolicy/util_linux_amd64.go

```diff
@@ -93,3 +139,421 @@ containerLoop:
+	var totalSize uint64
+	numaFreeMem := make(map[int]uint64)
+	for _, n := range numas {
+		freeMemSize, err := getNumaNodeFreeMemMB(systemNodeDir, n)
+		if err != nil {
+			return nil, fmt.Errorf("failed to getNumaNodeFreeMem for node %d, err %s", n, err)
+		}
+
+		numaFreeMem[n] = freeMemSize
+		totalSize += freeMemSize
+	}
+
+	// convert numa free memory ratio to a approximate value which is in the range [1, 10]
+	var fraction uint64 = 1
+	if totalSize > 10 {
+		fraction = (totalSize + 10) / 10
+	}
+
+	numaFreeMemRatio := make(map[int]int)
+	for numaID, freeMemSize := range numaFreeMem {
+		val := freeMemSize / fraction
+		if val == 0 {
+			continue
+		}
+		numaFreeMemRatio[numaID] = int(val)
+	}
+
+	return numaFreeMemRatio, nil
+}
+
+func getProcessPageStats(procDir string, pid int) (*smapsInfo, error) {
+	smapFile := filepath.Join(procDir, fmt.Sprintf("%d/smaps", pid))
+	lines, err := general.ReadFileIntoLines(smapFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to ReadLines(%s), err %v", smapFile, err)
+	}
+
+	info := &smapsInfo{}
+	var vma *vmaInfo
+	for _, line := range lines {
+		if vma == nil {
+			elems := strings.Fields(line)
+			if len(elems) == 0 {
+				continue
+			}
+
+			vmaRange := strings.Split(elems[0], "-")
+			if len(vmaRange) != 2 {
+				continue
+			}
+
+			vmaStart, err := strconv.ParseUint(vmaRange[0], 16, 64)
+			if err != nil {
+				continue
+			}
+
+			vmaEnd, err := strconv.ParseUint(vmaRange[1], 16, 64)
+			if err != nil {
+				continue
+			}
+
+			vma = &vmaInfo{
+				start:       vmaStart,
+				end:         vmaEnd,
+				virtualSize: -1,
+				rss:         -1,
+				pageSize:    -1,
+			}
+			continue
+		}
+
+		if strings.HasPrefix(line, "Size:") {
+			elems := strings.Fields(line)
+			if len(elems) != 3 {
+				continue
+			}
+
+			sizeKB, err := strconv.ParseInt(elems[1], 10, 64)
+			if err != nil {
+				continue
+			}
+			vma.virtualSize = sizeKB * 1024
+			continue
+		}
+
+		if strings.HasPrefix(line, "KernelPageSize:") {
+			elems := strings.Fields(line)
+			if len(elems) != 3 {
+				continue
+			}
+
+			pageSizeKB, err := strconv.ParseInt(elems[1], 10, 64)
+			if err != nil {
+				continue
+			}
+			vma.pageSize = pageSizeKB * 1024
+			continue
+		}
+
+		if strings.HasPrefix(line, "Rss:") {
+			elems := strings.Fields(line)
+			if len(elems) != 3 {
+				continue
+			}
+
+			rssKB, err := strconv.ParseInt(elems[1], 10, 64)
+			if err != nil {
+				continue
+			}
+			vma.rss = rssKB * 1024
+			continue
+		}
+
+		if strings.HasPrefix(line, "VmFlags:") {
+			if vma.virtualSize == -1 {
+				continue
+			}
+			if vma.rss == -1 {
+				continue
+			}
+			if vma.pageSize == -1 {
+				continue
+			}
+
+			info.totalRss += vma.rss
+			info.vmas = append(info.vmas, *vma)
+			vma = nil
+		}
+	}
+
+	return info, nil
+}
+
+func getProcessPagesNuma(pid int32, pagesCount uint64, pagesAddr []uint64) (pagesNuma []int32, err error) {
+	return movePages(pid, pagesCount, pagesAddr, nil)
```

Comment:
recommand to add the comment 

"nodes can also be NULL,  in  which  case  move_pages() does  not  move any pages but instead will return the node where each page currently resides, in the status array"

Reply:
done